lib/location/scan_cellular.c

```diff
@@ -99,23 +128,70 @@ int scan_cellular_start(struct k_sem *ncellmeas_ready)
 			scan_cellular_info.current_cell.id = modem_params.cell_id;
 			scan_cellular_info.current_cell.phys_cell_id = modem_params.phys_cell_id;
 		}
-		k_sem_give(scan_cellular_ready);
-		scan_cellular_ready = NULL;
+		goto end;
+	}
+	err = k_sem_take(&scan_cellular_sem_ncellmeas_evt, K_FOREVER);
+	if (err) {
+		/* Semaphore was reset so stop search procedure */
+		goto end;
 	}
+
+	/* Calculate the number of GCI cells to be requested. '-1' for current cell. */
+	scan_cellular_cell_count = scan_cellular_cell_count - 1 - scan_cellular_info.ncells_count;
+
+	if (scan_cellular_cell_count > 0) {
+
+		/* Initiate GCI search in a worker. GCI search cannot be initiated here
+		 * because previous NCELLMEAS hasn't been completed yet as giving semaphore
+		 * in lte_lc event handler takes the execution here too quickly.
+		 *
+		 * This function is running in location work queue so GCI search is
+		 * added to system work queue to avoid operation getting stuck.
```

Comment:
As discussed offline, GCI search can take a long time and it would be better not to execute it in the system work queue.

Reply:
It will just be triggered in system work queue by sending AT command so this would be fine.
However, it would be nice to be able to avoid use of system work queue here for code readability sake.

Reply:
Mutex implementation added that seemed to resolve these issues